### PR TITLE
[DiffusionGemma] Add foundation reference and QB2 bring-up harness

### LINKED ISSUE #47461: [DiffusionGemma] Bring up causal Gemma-4 26B-A4B backbone
**Part of #47452.** Milestone: foundation (prereq — de-risks ~80–90% of per-step compute).

Bring up the existing causal Gemma-4 26B-A4B backbone and validate it reproduces HF logits on the target HW **before** touching any diffusion-specific code.

### Scope
- Point `models/demos/gemma4/` at the DiffusionGemma checkpoint's text backbone weights.
- Validate logits PCC vs HF + coherent greedy generation on **QB2 (`P150x4`, 1×4)** — verified to fit + run (26B-A4B fits at TP=4, no OOM; experts TP-sharded). QB2/BHG enablement (memory budget + 4×8 TP) is split out to #47487.
- De-risk: MoE-128 / top-8 sparse path, shared MLP, final softcap, dual-θ RoPE, weight loading, max ctx 256K, multi-device TP.
- **Net-new weight module:** implement the **self-conditioning gated MLP** loader (no such module in `gemma4/tt/`). This gates the PCC harness (#47468).

### Reuse (already implemented)
- `models/demos/gemma4/tt/*` (model, attention, moe, router, experts, shared_mlp, rms_norm, ccl, precision) — **DONE**
- `tests/unit/test_model.py`, `tests/pcc_thresholds.json`
- **K=V tying is already in place** (not net-new): `attention_k_eq_v=True` is active on global layers of 12B/26B/31B and ties `v_w = k_w` (`tt/attention/weights.py:73`, `tt/model_config.py:45`). The bidirectional layers have no `v_proj`.
- **V is already scaleless-normed.** gemma4 norms V with `weight=None, with_scale=False` (`tt/attention/prefill.py:61`, `tt/attention/decode.py:84`), matching DiffusionGemma's `v_norm` (also `with_scale=False`).

### Verify vs HF before coding the diffusion delta
- Presence of `v_norm` (DiffusionGemma applies RMSNorm to Q/K/**V**). The real delta is a **per-layer reconciliation** of q/k/v-norm + K=V against `modeling_diffusion_gemma.py` (likely largely present) — not implementing a missing v_norm module.
- Router extra scales (`scale` / `scalar_root_size` / `per_expert_scale`).
- Whether encoder & decoder share weights.

### Acceptance
- Logits PCC ≥ threshold + coherent greedy generation on QB2.

### LINKED ISSUE #47468: [DiffusionGemma] Torch reference model and PCC harness
**Part of #47452.** Milestone: foundation (prereq).

Vendor the HF `DiffusionGemmaForBlockDiffusion` torch reference and build the accuracy harness that every later step validates against.

### Scope
- Vendor / wrap the HF reference under `models/experimental/diffusion_gemma/reference/`.
- Deterministic `generate` (fixed seed + fixed denoising schedule) → reproducible trajectory = the PCC oracle.
- PCC utilities for per-op, per-module, and **full-denoising-trajectory** comparison.
- **Extend PCC beyond logits.** A logit-PCC harness does not validate the diffusion *decisions*. Add checks for: **entropy** values (−Σ p·log p), **Gumbel-max argmax agreement**, and **multi-step denoise-trajectory fidelity** — especially under **bfp8** (small-probability drift can flip accept/remask and break convergence). There is **no entropy / −Σ p·log p computation anywhere** in gemma4 (the final-logit softcap `tanh(logits/30)*30` lives at `tt/model.py:718-722`), so the entropy harness is fully net-new (build on `ttnn.softmax`/`log`/`mul`/`sum`). Cross-ref #47472.
- **Determinism requires noise injection.** Token-for-token match is **not** achievable by regenerating Gumbel / random-renoise on device — on-device RNG won't reproduce torch's RNG bit-exactly, and entropy-budget accept/remask is noise-sensitive. The oracle harness must **inject the torch run's exact Gumbel noise + random renoise token ids** into the TT path; reserve regenerated-noise runs for distributional (not token-exact) checks.

### Reuse
- `tests/ttnn/utils_for_testing.py` (`assert_with_pcc`)
- `tech_reports/ttnn/comparison-mode.md` (auto per-op PCC)

### Acceptance
- **Reproducible oracle, version-PINNED.** Reference output reproducible under a fixed seed + fixed denoising schedule. The reference is reconciled against and pinned to **transformers 5.12.1** (the released build that ships `diffusion_gemma`; the vendored snapshot in `reference/_upstream.py` mirrors `main`@`49f6b78`). `tests/test_real_transformers_parity.py` asserts our primitives match the **installed** classes bit-for-bit and `test_upstream_parity.py` is the fallback for envs without `diffusion_gemma` — both must stay green across a transformers bump (the pin is the red line vs upstream drift; `main` is a moving target and is NOT the baseline).
- **Harness diffs a ttnn tensor vs torch at op / module / trajectory level, with explicit pass/fail bars:**
  - **op / module PCC ≥ 0.99** (measured: entropy bf16 0.999, self-conditioning 0.99996, bidirectional SDPA ≥ 0.99). Full causal-backbone logits **≥ 0.83** (26B-A4B, the Blackhole baseline) / ≥ 0.95 (12B dense).
  - **Decision red line — the bfp8 bar.** The decision-critical ops (per-token entropy, entropy-budget accept/remask, clean-argmax commit) run at **bf16 or fp32, never bfp8**. At bf16 vs the fp32 oracle: **accept-mask flip ≤ 0.5%** (measured 0.13%) and **commit-argmax flip ≤ 3%** (near-max ties; measured 1.3%) — enforced by `test_sampling.test_decision_dtype_red_lines_bf16_safe`. **bfp8 is rejected for decisions**: entropy PCC collapses to ~0.74 (measured on QB2) and `ttnn.argmax` refuses bfp8 — bfp8 is for bulk weight storage only. Under **injected** noise, the trajectory matches token-for-token (argmax / accept / canvas, 0 flips over ≤ 48 steps) at fp32/bf16.
- **3-phase oracle ownership (so no issue assumes the oracle lives elsewhere).** This harness owns the **denoise** (bidirectional) per-step decision oracle. The cross-phase **prefill (causal) → denoise → commit (causal)** KV-snapshot / activation oracle that #47474 (KV phase machine) and #47462 (bidirectional forward) validate against is produced from the real model via `hf_reference_generate()` (HF `generate()` exposes `past_key_values`): **#47474 owns snapshotting + validating those KV zones**; #47468 supplies the reference model and the denoise-decision diff.

### PR DESCRIPTION
**Pull request by zhenghaoniTT**
## Summary
- Add the DiffusionGemma reference layer, HF parity guards, sampling/trajectory PCC harnesses, and self-conditioning support.
- Add QB2 device validation coverage for backbone PCC, entropy/Gumbel decisions, bidirectional SDPA, self-conditioning, and 256K weights+KV memory budget.
- Document the QB2 memory budget and pinned DiffusionGemma test environment.

## Test plan
- `HF_MODEL=/home/zni/dg_models/gemma-4-26B-A4B-it pytest models/experimental/diffusion_gemma/tests -q`
- `DG_RUN_DEVICE=1 MESH_DEVICE=P150x4 HF_MODEL=/home/zni/dg_models/gemma-4-26B-A4B-it pytest models/experimental/diffusion_gemma/tests -q -s -k "not blackhole-1x1 and not blackhole-1x2" --tb=short --timeout=7200`

## Related issues
Part of #47452 (Support DiffusionGemma model). Milestone: **Foundation**.

This PR delivers the Foundation prerequisites and de-risks the Functional-core device work:

- **#47468** — Torch reference model + PCC harness: reference layer, HF bit-for-bit parity guard, sampling + decision-level trajectory PCC, self-conditioning. *Core deliverable of this PR.*
- **#47461** — Causal backbone PCC on QB2: DiffusionGemma ckpt at TP=4, logits PCC **0.877** (5-tok) / **0.847** (24-tok) ≥ 0.83 baseline. *Measured here.*
- **#47487** — QB2 HW enablement: weights+KV memory budget + pinned test env. *Partial* — QB2 weights+KV fit + budget land here; **256K e2e prefill chunking and Galaxy 4×8 remain open**.

De-risk spikes for the Functional-core issues land here too — **isolated, validated on QB2, not yet integrated into the model forward**:
- **#47462** — bidirectional canvas SDPA (mask geometry + non-causal SDPA path), 4/4.
- **#47463** — entropy-budget acceptance chain (`sort → cumsum → exclusive-prefix → scatter`), 5/5.
- **#47472** — on-device entropy / Gumbel-max sampling primitives.

> Not using `Closes` keywords — #47487 is partial and the spike issues track their (un-started) integration. Convert to `Closes #47468` / `Closes #47461` if you want those to auto-close on merge.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:zni/diffusion-gemma-foundation-draft)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:zni/diffusion-gemma-foundation-draft)

### CODE REVIEW

File: models/experimental/diffusion_gemma/requirements.full.lock.txt [hunk 1]
```diff
@@ -0,0 +1,365 @@
+# DiffusionGemma — FULL env snapshot (exact dist-info), 2026-06-24
```
Comment (williamlyTT) [inline models/experimental/diffusion_gemma/requirements.full.lock.txt]: Is this file expected to be checked in?

**Review by williamlyTT — APPROVED** [review]
approving to unblock but please check the comment

File: models/experimental/diffusion_gemma/requirements.full.lock.txt (hunk 1 above)
Comment (zhenghaoniTT) [reply models/experimental/diffusion_gemma/requirements.full.lock.txt]: It has been removed. Thanks for the review

**Review by mtairum — APPROVED** [review]
